### ISSUE TITLE
Parts of nicknames recognized as coordinates

### ISSUE THREAD

**Issue by Sofia-mal** [online-go/online-go.com#2026]
https://online-go.com/game/44531856

![immagine](https://user-images.githubusercontent.com/100384001/194497262-e6f6e88c-c140-4167-a554-e4df53b09d63.png)

Not an important issue at all, but I think [P1.04](https://online-go.com/player/1039370/P1.04)'s nickname is not a reference to P1 on the board and it should not be recognized as a coordinate.

**Comment by kobakitty** [comment]
Same thing happens when mentioning users with usernames like "player 55" or "user 1213" - it got turned into link for players who have the corresponding user-id

**Comment by Sofia-mal** [comment]
I don't think it's the same thing and what you are saying is probably intended :)

**Comment by Jlll1** [comment]
I'd like to work on this. How would we like to see this resolved? I'd go with making the user be implicit about embedding anything into his message. E. g. instead of converting `P1` or `player 55` into a link by default, require the user to wrap the link, like `[[P1]]`, `((P1))`, `[[player 55]]` etc.

**Comment by benjaminpjones** [comment]
TBH I don't think this is a great issue to start with - I think the system messages are generated on the backend which isn't open source.

As far as ideas go - there are actually player escapes already. See [chat_markup.tsx](https://github.com/online-go/online-go.com/blob/421b711821d74817cbbcba39b22dda5ee8a7596a/src/components/Chat/chat_markup.tsx#L122-L161).  It has a few ways to add a player such as `%%%PLAYER-12345%%%` and `@"benjito"`.  We could probably have the server bot send one of these instead of the plaintext username.

@GreenAsJade any thoughts?

**Comment by anoek** [comment]
I think this should be all front end, and so a good first issue to work on?

I believe the relevant code is here: https://github.com/online-go/online-go.com/blob/devel/src/views/Game/GameChat.tsx#L595-L616

**Comment by benjaminpjones** [comment]
I thought the solution was to change what `system [9d]` adds to chat. But I think I may have read that wrong.

I see now that the proposed solution is to wrap moves like `[[M1]]`.  It seems like that will reduce discoverability, but I'm not against it.

**Comment by GreenAsJade** [comment]
As a side note, we can of course easily change what `system [9d]` says, though it is indeed on the back end.

An also side note:

> Same thing happens when mentioning users with usernames like "player 55" or "user 1213" - it gets turned into link for players who have the corresponding user-id

... yes that's what is supposed to happen :)    It's just that in this case the markup is mistaking a user reference for a position reference.

I agree with @benjito that the discoverability of this is fairly cool.

I also agree that  `@player` would seem the ideal markup choice.   That at least is mildly discoverable, in that folk will type that for talking about a person in other contexts?

If we go with that then @jlll1  could take a crack at it, and look into the precedence issue on the way, then I can update the server.

**Comment by Jlll1** [comment]
I agree that it's very convenient to have the linking done by default. Whenever you're typing a message with things like `tournament ...`, `player ...`, `game ...`, you most likely want the link to be there. The only real edge case is when the player nicknames contain these keywords, like in @Sofia-mal 's example. If that is true, then I guess there could be better solutions for that problem than requiring explicit syntax for linking.

**Comment by GreenAsJade** [comment]
... yeah, the question is "what" 😄 

FWIW there's another markup issue, if you're keen: https://github.com/online-go/online-go.com/issues/2021 🤣

**Comment by shinuito** [comment]
> I see now that the proposed solution is to require markup to be explicit, so moves would look like `[[M1]]`. It seems like that will reduce discoverability, but I'm not against it.

I’m against it :) It’s very nice to be able to just write P1 and have it turn into a coordinate. Not everyone might be using the same coordinates and it’s great to have the hover over highlight the coordinate. 

I’d rather escape usernames when they’re meant rather than having extra syntax for coordinates since surely coordinates are far more common. I’ve seen people even share lists of coordinates as variations to discuss rather than the analyse + share button :)

**Comment by anoek** [comment]
Yeah I don't think we want to make it hard to type coordinates. In the case of P1.04 I would say the fix would be to fix the pattern matching to ensure there's nothing trailing perdiods without whitespace or something. For player 2342 kind of things, I should probably just restrict that from being a username folks can register.

**Comment by BHydden** [comment]
Yeah I was going to say I think the problem is more with players usernames rather than with the markup.

**Comment by benjaminpjones** [comment]
>In the case of P1.04 I would say the fix would be to fix the pattern matching to ensure there's nothing trailing perdiods without whitespace or something.

A whitespace check would fix that specific case, but usernames like **I love A1 sauce** are valid and I don't think we want to change that.

**Comment by bhostetler18** [comment]
Are there any cases where the desired behavior is a special markup formatting inside a substring that already has a different markup replacement?

If not, a more robust (but slightly more complicated) solution might be to modify how the chat markup processes each string. Currently, it goes through the original string with the first replacement type and splits the string into matching (and non-matching) sections, then applies all subsequent replacements to _each and every_ section generated thus far. If a matching section were instead marked as “handled” when found and excluded from future replacements, it would prevent this issue and other similar ones. It would also automatically introduce precedence determined by the order of the entries in the replacements array.

**Comment by benjaminpjones** [comment]
Another example of an erroneous "move":

```
https://ogs-forums.s3.dualstack.us-east-1.amazonaws.com
```

The s3 gets the move markdown treatment:

<img width="310" alt="Screen Shot 2022-11-29 at 12 38 04 PM" src="https://user-images.githubusercontent.com/25233703/204642577-694fb835-f57c-45e9-b62b-82a952d7e085.png">

**Comment by DimensionalDot** [comment]
Looking back at this, it seems to have been resolved. I believe by #3370 or #3371. Should be closable. @anoek @GreenAsJade
